Added: callback for releasing clicks on labels

dicene opened the pull request:
Adds setLabelReleaseCallback to the lua interpreter. (Mudlet Code)

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -3208,6 +3208,56 @@ int TLuaInterpreter::setLabelClickCallback( lua_State *L )
+    Host * pHost = TLuaInterpreter::luaInterpreterMap[L];
+    QString text(luaSendText.c_str());
+    QString name(luaName.c_str());
+    mudlet::self()->setLabelReleaseCallback( pHost, text, name, pE );
```
Not sure this method exists in the `mudlet` class, and unfortunately this will need revision to the `TLabel` class to support the arguments that the `TEvent` instance the `pE` points to.  I suspect that the `TLabel` storage in Mudlet Maps may be affected by such a change as well. 😞

dicene replied:
I removed my own comment somehow? :suspect:

I added the changes to TLabel in the second commit, and the method was added to the mudlet class in the first commit. Never really looked at any of the map stuff, so I have no clue if my changes to TLabel will break things there.

SlySven commented:
@dicene 
Don't forget to update the copyright header with a line for yourself at the top of the c++ files you modify! :grin:

Have you checked that the release related `TEvent` parameters in `TLabel` i.e. `TLabel::mRelease` and `TLabel::mrParameters` are stored between sessions?  I suspect that this will need changes to both the profile saving/loading code and (as the MAP can also have labels) may also need changes to the map saving/loading stuff as well...

dicene commented on src/TLabel.h:
```diff
@@ -43,17 +43,21 @@ Q_OBJECT
 QString           mEnter;
 QString           mLeave;
 TEvent            mpParameters;
+TEvent            mrParameters;
```
Oh, heh. This is what I mentioned on Gitter. I'm not very hip to naming conventions yet. :grimacing: I guess it's something I'll pick up over time. I can go through and fix every reference to mClickParameters in a commit so it's nice and clean everywhere.

dicene replied:
I actually don't see any references to it outside of the TLabel. I'll give it a test and see if it works replaced with your name ideas.

SlySven commented on src/mudlet.cpp:
```diff
@@ -1257,7 +1257,19 @@ bool mudlet::setLabelClickCallback( Host * pHost, const QString & name, const QS
     if( labelMap.contains( name ) )
     {
-        labelMap[name]->setScript( pHost, func, pA );
+        labelMap[name]->setClick( pHost, func, pA );
```
Would it be reasonable to actually pass the return value from the `TLabel::setClick(...)` (and the other setRelease() and probably the enter and leave things as well) to the user?  It indicates (with a true value) that the label was found for the Lua "script" to be allocated to the event IIRC and it would make sense and be consistent to do so, IMHO.

dicene replied:
I would also have to modify the setClick function itself to have it return something to get this working, yeah? 
The current setClick is a void, so I'm guessing returning it without modifying the code would just have it returning NULL?
void TLabel::setClick( Host * pHost, const QString & func, const TEvent & args )
{
    mpHost = pHost;
    mClick = func;
    mClickParams = args;
}

SlySven replied:
Note, I heavily edited my previous comment as I thought the `TLabel` methods had return values - I was wrong and I re-targetted my comments to the `mudlet` methods so that the `TLuaInterpreter` ones return the true or false from the `mudlet` ones instead of blindly returning true whatever the `mudlet` methods return.

vadi2 commented:
I've updated your original comment @dicene, added a little list to keep track of changes needed.

vadi2 commented:
Where are we at with the Mudlet code - is that ready to be tested?

dicene commented:
Mudlet code is finished. Worked properly when I tested it.

vadi2 commented:
OK, will have a look as well.

On Wed, 7 Dec 2016 11:48 am dicene, <notifications@github.com> wrote:

> Mudlet code is finished. Worked properly when I tested it.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Mudlet/Mudlet/pull/355#issuecomment-265331290>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGxjDGipxvzDuzEvTZBmTMZcmMEbCTvks5rFhBSgaJpZM4LDa9b>
> .
>

SlySven commented:
Um, don't forget to add your Copyright details to the top of the source files you have changed.  It makes it clear that your changes are released under the GPL2.1+ terms we need...!

dicene commented:
I'll go back through all of these little PRs to fix that hopefully tonight. In one of those programming moods! :honeybee:

SlySven commented on src/dlgTriggerEditor.cpp:
```diff
@@ -361,6 +362,7 @@ dlgTriggerEditor::dlgTriggerEditor( Host * pH )
 
     QAction * profileSaveAction = new QAction( QIcon ( QStringLiteral( ":/icons/document-save-all.png" ) ), tr("Save Profile"), this);
     profileSaveAction->setEnabled( true );
+    profileSaveAction->setShortcut(tr("Ctrl+Shift+S"));
```
As above.

dicene commented:
Sorry. Was up way to late working on things last night, and I guess the fatigue was making me sloppier than normal.

Added the profile save stuff in https://github.com/Mudlet/Mudlet/pull/355/commits/be51d58029814400297a211e632abd5d147f9161

Removed it in https://github.com/Mudlet/Mudlet/pull/355/commits/97fee633f3bfcf186a24df44e314b627a5c21f3e

SlySven commented:
I am happy with this now. :+1:

dicene commented:
If I'm not mistaken, the official up-to-date version of both Geyser and Vyzor are located outside of this Repo anyway, so I guess there's no way for me to include those changes on this PR either way, heh. Yeah. I'll remove them from the task list at the beginning and this should be good to merge then.

To confirm, the official Geyser would be: https://github.com/vadi2/mudlet-lua
No clue where the Vyzor stuff is...

keneanung commented:
Vyzor should be at https://github.com/Oneymus/Vyzor

vadi2 commented:
Thanks everyone for reviewing and @dicene for the improvement!

SlySven commented:
@vadi2 \<whisper>Do you think we've got dicene properly hooked into Mudlet development yet? 😜 \</whisper>

vadi2 commented:
Hopefully being able to make changes in your favourite client and then actually using them is fulfulling!

dicene commented:
Heh, very fulfilling. If only the rest of the software I use in my life(poorly coded systems I'm forced to utilize at work, Android(on a phone with no documented root procedure), my Credit Union's mobile app, etc), was like this so I could make improvements that I'd enjoy throughout the rest of my day. =/

Future projects:

- Finish up the labelReleaseCallback stuff for Geyser(which I think I have working, but I need some clarification to questions I asked on the PR I've got on that Repo) and Vyzor(which just looks weird to me).
- Make the scrollback act more intuitively when you use the mouse wheel or page+down to scroll to the bottom of the scrollback.
- Enhance the main console so that clicking and click-dragging over timestamps selects the entire line or lines, which was a solid boost to my productivity in CMUD when doing things like adding triggers while scrolling back through a long fight in the game I play.
- Possibly adding a way to change the default link color, since I find the dark blue color very difficult to read against a black background.

SlySven commented:
> Spending more time in the past several months working on Mudlet than using it, and I don't see that changing particularly soon...

Join the club...! :wink:
